### PR TITLE
ensure childModelData index is in sync with trackedarray index #162

### PR DESCRIPTION
**Pull request by dnalagatla**

### CODE REVIEW

File: addon/m3-tracked-array.js [hunk 1]
```diff
@@ -25,20 +25,35 @@ export default class M3TrackedArray extends ArrayProxy {
   }
 
   replaceContent(idx, removeAmt, newItems) {
-    newItems = newItems.map((item, idx) => {
+    // Update childModelDatas array
+    // mapping to array of nested models
+    const childModelDatas = this._model._internalModel._modelData.__childModelDatas;
+    const key = this._key;
+    if (childModelDatas && childModelDatas[key] && Array.isArray(childModelDatas[key])) {
```
Comment (hjdivad) [inline addon/m3-tracked-array.js]: This array check should probably be an assertion

File: addon/m3-tracked-array.js [hunk 2]
```diff
@@ -25,20 +25,35 @@ export default class M3TrackedArray extends ArrayProxy {
...
+      const newItemsInChildModelData = new Array(newItems.length);
+      Array.prototype.splice.apply(
+        childModelDatas[key],
+        [idx, removeAmt].concat(newItemsInChildModelData)
+      );
+    }
+
+    let startIdx = idx;
+    newItems = newItems.map(item => {
       if (isResolvedValue(item)) {
```
Comment (hjdivad) [inline addon/m3-tracked-array.js]: I think we should merge this pr without resolving this issue as it's an edge case that we don't actually handle

BUT

Can we please log a bug to remind ourselves to test this case; in particular, to test that inserting a resolved value into a tracked array (and especially at some position other than the start or end) leaves everything in a correct state

File: addon/m3-tracked-array.js [hunk 3]
```diff
@@ -25,20 +25,35 @@ export default class M3TrackedArray extends ArrayProxy {
...
+    // mapping to array of nested models
+    const childModelDatas = this._model._internalModel._modelData.__childModelDatas;
+    const key = this._key;
+    if (childModelDatas && childModelDatas[key] && Array.isArray(childModelDatas[key])) {
+      const newItemsInChildModelData = new Array(newItems.length);
+      Array.prototype.splice.apply(
+        childModelDatas[key],
+        [idx, removeAmt].concat(newItemsInChildModelData)
+      );
+    }
```
Comment (hjdivad) [inline addon/m3-tracked-array.js]: I think it's confusing to have tracked array know so much about the internals of modeldata.

I suspect there's something better we can do, but for now let's do something like add `modelData._resizeChildModelData(idx,removeLength,addLength)`  so model data can own the logic here rather than tracking array know about its internals

File: addon/m3-tracked-array.js [hunk 4]
```diff
@@ -25,20 +25,35 @@ export default class M3TrackedArray extends ArrayProxy {
...
+
+    let startIdx = idx;
+    newItems = newItems.map(item => {
       if (isResolvedValue(item)) {
         associateRecordWithRecordArray(item, this);
         return item;
       }
 
+      const index = startIdx;
+      startIdx++;
```
Comment (hjdivad) [inline addon/m3-tracked-array.js]: i would personally see this more clearly if the index were computed as the sum of `idx` and the index argument to `map`

File: addon/m3-tracked-array.js (hunk 1 above)
Comment (dnalagatla) [reply addon/m3-tracked-array.js]: Added an assertion.

Comment (dnalagatla) [reply addon/m3-tracked-array.js]: Added an assertion.

File: addon/m3-tracked-array.js (hunk 3 above)
Comment (dnalagatla) [reply addon/m3-tracked-array.js]: Moved the implementation to `modelData._resizeChildModelData`

File: addon/m3-tracked-array.js (hunk 4 above)
Comment (dnalagatla) [reply addon/m3-tracked-array.js]: updated the code to compute the index.
